src/binder_ims.c

```diff
@@ -289,6 +289,7 @@ binder_ims_probe(
             self->caps |= OFONO_IMS_VOICE_CAPABLE;
         }
     }
+    self->reg->caps = self->caps;
```

Comment:
Is there a way to avoid making multiple copies of the same thing?

Reply:
Would it be ugly to move the caps to BinderImsReg and assign/access that from BinderIms, that way it wouldn't need to duplicate the values.

Reply:
Sounds ugly, since `BinderImsReg` was supposed to just keep track of the IMS registration status, and the caps is a plugin-wide thing that doesn't depend on the registration status.

Reply:
Would it be ok to replace the uses for BinderImsReg with BinderIms (struct could still not be public) and maybe expose two functions one for registered state and another for caps? That way BinderImsReg wouldn't need to have new variable.

Reply:
But it seems that way would add quite a lot of code, BinderIms would have to become GObject so we can manage references etc.

Reply:
Perhaps it's fine the way you've done it. Anything I could think of, was significantly more sophisticated, and didn't feel like it was worth the trouble. It's an internal interface after all, it can be changed later when/if somebody has an idea how to do it better.

Reply:
I now pushed a version which implements that way. I also added another commit which fixes another issue related to ims.

Reply:
@monich let me know if the new code I pushed is too complex for fixing this issue

Reply:
@monich  We need some acceptable way of getting these caps working. This is blocking an ims plugin at the moment.

Reply:
looking at it...

Reply:
The changes overall make sense to me, except that the code can be simplified a bit. AFAICT `BinderIms` doesn't need to be a `GObject`. And even if it is a `GObject`, it doesn't need to be derived from `BinderBase` (which is a reusable mechanism for queuing property change signals) and doesn't need to be split to public and private parts (since all its data structures are private, i.e. defined and used inside a single .c file).

`GObject` is a relatively heavy thing with significant amount of boilerplate code (albeit most of it is hidden behind macros) and certain runtime overhead. It's useful primarily as a signal emission mechanism. `BinderIms` doesn't emit signals.

Typically, I use `GObject` when I need its signal emission/handling mechanism, and I choose a simpler approach for simple refcounting e.g. https://github.com/sailfishos/libglibutil/blob/master/src/gutil_weakref.c

Reply:
That does seem like a better way to do refcounting. I'll have a look.

Reply:
BTW I merged PR #31. I suggest to rebase this one.

Reply:
I pushed a new version with only simple int based ref counting similar to what gutil_weakref does internally. I used this way because gutil_weakref requires the pointer to be GObject so just using int was simpler.